File: [84] Largest Rectangle in Histogram/84.largest-rectangle-in-histogram.java
import java.util.Deque;

/*
 * @lc app=leetcode id=84 lang=java
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
class Solution {
    public int largestRectangleArea(int[] heights) {
        // My first solution
        if (heights.length == 1) return heights[0];
        int n = heights.length;
        int left = -1, right = 0;
        while (right < n && heights[right] >= heights[0]) right++;
        int area = heights[0]*(right-left-1);
        for (int i = 1; i < n; i++) {
            if (heights[i] > heights[i-1]) {
                left = i-1;
                right = i+1;
            }
            while (left >=0 && heights[left] >= heights[i]) left--;
            while (right < n && heights[right] >= heights[i]) right++;
            area = Math.max(area, heights[i]*(right-left-1));
        }
        return area;
    }
}
// @lc code=end

File: [84] Largest Rectangle in Histogram/84.largest-rectangle-in-histogram.cpp
/*
 * @lc app=leetcode id=84 lang=cpp
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        // stack solution
        // https://www.youtube.com/watch?v=VNbkzsnllsU
        int n = heights.size();
        if (n == 1) return heights[0];
        stack<int> st;    // record the index
        int area = 0;
        for (int i = 0; i <= n; i++) {
            int h = (i == n ? 0 : heights[i]);
            while (!st.empty() && h < heights[st.top()]) {
                int curHeight = heights[st.top()];
                st.pop();
                int rightBoundary = i - 1;
                int leftBoundary = st.empty() ? 0 : st.top() + 1;
                int width = rightBoundary - leftBoundary + 1;
                area = max(area, (curHeight * width));
            }
            st.push(i);
        }
        return area;
    }
};
// @lc code=end

